plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -0,0 +1,568 @@
+            cause = Builtin () $ DynBuiltinName dynName
+        Just mean -> pure mean
+
+-- | The computing part of the CEK machine.
+-- Either
+-- 1. adds a frame to the context and calls 'computeCek' ('Force', 'Apply')
+-- 2. calls 'returnCek' on values ('Delay', 'LamAbs', 'Constant', 'Builtin')
+-- 3. returns 'EvaluationFailure' ('Error')
+-- 4. looks up a variable in the environment and calls 'returnCek' ('Var')
+
+computeCek
+    :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
+    => Context uni -> CekValEnv uni -> TermWithMem uni -> CekM uni (Term Name uni ())
+-- s ; ρ ▻ {L A}  ↦ s , {_ A} ; ρ ▻ L
+computeCek ctx env (Force _ body) = do
+    spendBudget BForce (ExBudget 1 1) -- TODO
+    computeCek (FrameForce : ctx) env body
+-- s ; ρ ▻ [L M]  ↦  s , [_ (M,ρ)]  ; ρ ▻ L
+computeCek ctx env (Apply _ fun arg) = do
+    spendBudget BApply (ExBudget 1 1) -- TODO
+    computeCek (FrameApplyArg env arg : ctx) env fun
+-- s ; ρ ▻ abs α L  ↦  s ◅ abs α (L , ρ)
+computeCek ctx env (Delay ex body) =
+    -- TODO: budget?
+    returnCek ctx (VDelay ex body env)
+-- s ; ρ ▻ lam x L  ↦  s ◅ lam x (L , ρ)
+computeCek ctx env (LamAbs ex name body) =
+    -- TODO: budget?
+    returnCek ctx (VLamAbs ex name body env)
+-- s ; ρ ▻ con c  ↦  s ◅ con c
+computeCek ctx _ (Constant ex val) =
+    returnCek ctx (VCon ex val)
+-- s ; ρ ▻ builtin bn  ↦  s ◅ builtin bn arity arity [] [] ρ
+computeCek ctx env (Builtin ex bn) = do
+    -- TODO: budget?
+  arity <- arityOf bn
+  returnCek ctx (VBuiltin ex bn arity arity 0 [] env)
+-- s ; ρ ▻ error A  ↦  <> A
+computeCek _ _ (Error _) =
+    throwingWithCause _EvaluationError (UserEvaluationError CekEvaluationFailure) . Just $ Error ()
+-- s ; ρ ▻ x  ↦  s ◅ ρ[ x ]
+computeCek ctx env (Var _ varName) = do
+    spendBudget BVar (ExBudget 1 1) -- TODO
+    val <- lookupVarName varName env
+    returnCek ctx val
+
+-- | Call 'dischargeCekValue' over the received 'CekVal' and feed the resulting 'Term' to
+-- 'throwingWithCause' as the cause of the failure.
+throwingDischarged
+    :: ( MonadError (ErrorWithCause e (Term Name uni ())) m
+       , Closed uni, uni `Everywhere` ExMemoryUsage
+       )
+    => AReview e t -> t -> CekValue uni -> m x
+throwingDischarged l t = throwingWithCause l t . Just . void . dischargeCekValue
+
+-- | The returning phase of the CEK machine.
+-- Returns 'EvaluationSuccess' in case the context is empty, otherwise pops up one frame
+-- from the context and uses it to decide how to proceed with the current value v.
+--  'FrameForce': call forceEvaluate.  If v is a lambda then compute the body of v;
+--     if v is a builtin application then check that it's expecting a type argument,
+--     either apply the builtin to its arguments and return the result,
+--     or extend the value with @force@ and call returnCek;
+--     if v is anything else, fail.
+--  'FrameApplyArg': call applyEvaluate. If v is a lambda then discard the type
```

Comment:
This section seems wrong: we just pop `FrameApplyFun` on the stack and compute the argument.

Reply:
I think we could leave the description of what `applyEvaluate` does to the comment of `applyEvalute`, this applies to the typed CEK machine too.